control/passivity.py

```diff
@@ -45,43 +84,127 @@ def ispassive(sys):
+            n, eps, 1.0, make_LMI_matrix)
+        sys_constants = -make_LMI_matrix(np.zeros_like(A), rho, eps, 1.0)
+    else:
+        sys_matrix_list = _make_P_basis_matrices(
+            n, 1.0, eps, make_LMI_matrix)
+        sys_constants = -make_LMI_matrix(np.zeros_like(A), eps, nu, 1.0)
 
-    matrix_list = []
-    state_space_size = sys.nstates
-    for i in range(0, state_space_size):
-        for j in range(0, state_space_size):
-            if j <= i:
-                P = np.zeros_like(A)
-                P[i, j] = 1.0
-                P[j, i] = 1.0
-                matrix_list.append(make_LMI_matrix(P).flatten())
+    sys_coefficents = np.vstack(sys_matrix_list).T
 
-    coefficents = np.vstack(matrix_list).T
-
-    constants = -np.vstack((
-        np.hstack((np.zeros_like(A),  - C.T)),
-        np.hstack((- C, -D - D.T)))
-    )
+    # LMI to ensure P is positive definite
+    P_matrix_list = _P_pos_def_constraint(n)
+    P_coefficents = np.vstack(P_matrix_list).T
+    P_constants = np.zeros((n, n))
 
+    # cost function
     number_of_opt_vars = int(
-        (state_space_size**2-state_space_size)/2 + state_space_size)
+        (n**2-n)/2 + n)
     c = cvx.matrix(0.0, (number_of_opt_vars, 1))
+    c = cvx.matrix(np.append(np.array(c), -1.0))
+
+    Gs = [cvx.matrix(sys_coefficents)] + [cvx.matrix(P_coefficents)]
+    hs = [cvx.matrix(sys_constants)] + [cvx.matrix(P_constants)]
 
     # crunch feasibility solution
     cvx.solvers.options['show_progress'] = False
-    sol = cvx.solvers.sdp(c,
-                          Gs=[cvx.matrix(coefficents)],
-                          hs=[cvx.matrix(constants)])
+    sol = cvx.solvers.sdp(c, Gs=Gs, hs=hs)
+    return sol["x"]
+
 
-    return (sol["x"] is not None)
+def get_output_fb_index(sys):
+    '''
+    Returns the output feedback passivity index for the input system. This is largest gain that can be placed in
+    positive feedback with a system such that the new interconnected system is passive.
+    Parameters
+    ----------
+    sys: An LTI system
+        System to be checked.
+
+    Returns
+    -------
+    float: 
+        The OFP index 
+    '''
+    sol = _solve_passivity_LMI(sys, nu=eps)
+    if sol is None:
+        return -np.inf
+    else:
+        return sol[-1]
+
+
+def get_input_ff_index(sys, index_type=None):
+    '''
+    Returns the input feedforward passivity (IFP) index for the input system. This is the largest gain that can be 
+    placed in negative parallel interconnection with a system such that the new interconnected system is passive.
+    Parameters
+    ----------
+    sys: An LTI system
+        System to be checked.
+
+    Returns
+    -------
+    float: 
+        The IFP index 
+    '''
+
+    sol = _solve_passivity_LMI(sys, rho=eps)
+    if sol is None:
+        return -np.inf
+    else:
+        return sol[-1]
+
+
+def ispassive(sys, rho=None, nu=None):
+    '''
+    Indicates if a linear time invariant (LTI) system is passive
+
+    Parameters
```

Comment:
Lacks the description for `rho` and `nu`

Reply:
meant to remove them